sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledSymmetricHashJoinExec.scala

```diff
@@ -0,0 +1,1105 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.{mutable, BitSet}
+
+import ai.rapids.cudf.{ContiguousTable, HostMemoryBuffer}
+import ai.rapids.cudf.JCudfSerialization.SerializedTableHeader
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource}
+import com.nvidia.spark.rapids.GpuMetric._
+import com.nvidia.spark.rapids.GpuShuffledSymmetricHashJoinExec.JoinInfo
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.RmmRapidsRetryIterator.withRetryNoSplit
+import com.nvidia.spark.rapids.ScalableTaskCompletion.onTaskCompletion
+import com.nvidia.spark.rapids.shims.{GpuHashPartitioning, ShimBinaryExecNode}
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.catalyst.plans.Inner
+import org.apache.spark.sql.catalyst.plans.physical.Distribution
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec
+import org.apache.spark.sql.rapids.execution.{ConditionalHashJoinIterator, GpuCustomShuffleReaderExec, GpuHashJoin, GpuShuffleExchangeExecBase, HashJoinIterator}
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+object GpuShuffledSymmetricHashJoinExec {
+  /** Utility class to track bound expressions and expression metadata related to a join. */
+  case class BoundJoinExprs(
+      boundBuildKeys: Seq[GpuExpression],
+      buildTypes: Array[DataType],
+      boundStreamKeys: Seq[GpuExpression],
+      streamTypes: Array[DataType],
+      streamOutput: Seq[Attribute],
+      boundCondition: Option[GpuExpression],
+      numFirstConditionTableColumns: Int,
+      compareNullsEqual: Boolean,
+      buildSideNeedsNullFilter: Boolean)
+
+  object BoundJoinExprs {
+    def bind(
```

Comment:
nit: A comment here about how this is called after the build side is determined would be nice.

Reply:
Will do, although the build side is a parameter to this method so it would be quite the trick to call it before that since GpuShuffledSymmetricHashJoinExec doesn't start with a GpuBuildSide specified.

Reply:
Yes I totally get that, but it is different from the other join code so it threw me a bit when I looked at it first. That is why it is just a nit.